include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
 	}
 }
 
-void flipCurrInteraction()
-{
-	currInteraction.isotropic.N = -currInteraction.isotropic.N;
-	currInteraction.isotropic.NdotV = -currInteraction.isotropic.NdotV;
-	currInteraction.T = -currInteraction.T;
-	currInteraction.B = -currInteraction.B;
-	currInteraction.TdotV = -currInteraction.TdotV;
-	currInteraction.BdotV = -currInteraction.BdotV;
-}
-//call before executing an instruction/evaluating bsdf
-void handleTwosided_interactionOnly(inout bool ts_flag, in instr_t instr)
+#ifdef GEN_CHOICE_STREAM
+void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache _microfacet, in bool skip)
 {
-#ifndef NO_TWOSIDED
-	if (!ts_flag && instr_getTwosided(instr))
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons) [(is_bxdf && is_bsdf) actually tells that op is bsdf]
+	// (is_bxdf && is_bsdf) -> BSDF
+	// (is_bxdf && !is_bsdf) -> BRDF
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+
+	uvec3 regs = instr_decodeRegisters(instr);
+	params_t params;
+	bsdf_data_t bsdf_data;
+	mat2x3 ior;
+	mat2x3 ior2;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+
+	const bool run = !skip;
+
+	if (is_bxdf_or_combiner && run)
 	{
-		ts_flag = true;
-		if (currInteraction.isotropic.NdotV<0.0)
-		{
-			flipCurrInteraction();
-		}
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
 	}
-#ifdef OP_BUMPMAP
-	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
-#endif
-#endif	
-}
-//call before executing an instruction/evaluating bsdf
-void handleTwosided(inout bool ts_flag, in instr_t instr)
-{
-#ifndef NO_TWOSIDED
-	if (!ts_flag && instr_getTwosided(instr))
+
+	vec3 eval = vec3(0.0);
+	float pdf = 0.0;
+
+	if (is_bxdf && run)
 	{
-		ts_flag = true;
-		if (currInteraction.isotropic.NdotV<0.0)
+		//speculative execution
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
+#ifdef ALL_ISOTROPIC_BXDFS
+		float one_minus_a2 = 1.0 - a2;
+#else
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
+#endif
+
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
+		const vec3 albedo = params_getReflectance(params);
+
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
```

Comment:
this one you can hoist and add the condition `>FLT_MIN` to `is_bxdf && run` because VNDF sampling (VNDF sampling can produce L samples that are shadowed, but cannot produce V samples that are masked)

Reply:
and obvs initialize `is_valid=true` after the hoist